### DISCUSSION TITLE
Pin defaults on boot-up.

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11193] [General]
I haven't found any reference in the micropython documentation that refers to pin defaults on boot-up.
Also, Espressif is very sparse with their documentation; they don't provide any schematics of, as a minimum, the typical I/O block.

Examples:
- For p1 = Pin(1); what are the defaults on boot-up?
- For p1 = Pin(1, Pin.OUT); what's the default output mode?
  Also, if the default output mode is configured as "(normal) output" [from micropython-glenn, class Pin]; does "normal" mean active Tri-State?

Thanks, MM

**Comment by scruss** [comment]
There are only a couple of pins that are in a reliable state on the ESP8266 at boot. Actual behaviour was discovered empirically in references such as [ESP8266 GPIO Behaviour at Boot](https://rabbithole.wwwdotorg.org/2017/03/28/esp8266-gpio.html)

**Comment by mmliam** [comment]
scruss,
Thanks for the information. It's unfortunate that Espressif provides such incomplete information on their products. It's particularly frustrating when one is familiar with the voluminous information provided by companies such as Intel, or Microchip. 
This some information I was able to locate: https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/

**Reply by scruss** [reply]
Espressif produce cheap wifi chipsets for the very lowest-end tablets. The fact they're usable as microcontrollers was only discovered by hobbyists in the last decade. Espressif's open source software/documentation (and financial support for MicroPython) has been incredible for such a small company.

As for Microchip, I still can't do anything with PIC development on Linux without forking over money. Intel's rapid exit from the IoT hobbyist market (Edison: 2014-2017) left a lot of people with stranded hardware.

**Comment by mmliam** [comment]
I do agree that hardware support for Espressif's products, and the easy ability to re-code Espressif products to micropython (from  the AT commands or Lua), with tools such as Thonny is exceptional. Not sure what you mean by financial support for micropython?

However, I still maintain Espressif's documentation is minimal, compared to Intel or Microchip. For example you may have seen my post #11196 regarding a vague reference to the ESP8266 being able to sample it's own supply voltage via the TOUT (and external ADC).

According to the ESP8266EX Data sheet

Pin 6, TOUT
ADC pin. It can be used to test the power-supply voltage of
VDD3P3 (Pin3 and Pin4) and the input power voltage of TOUT
(Pin 6). However, these two functions cannot be used
simultaneously.

The way I read the spec, there are two sampling functions, internal supply, or external via ADC. However, I have not been able to find any information on Espressif demonstrating how to implement testing VDD3P3 (Pin3 and Pin4). I have implemented an external voltage divider to sample the supply voltage, but it would sure simplify things, for a legacy device, if the ESP8266 already had an internal voltage divider and the sampling could be accomplished solely in software, and further does micropython the voltage supply sampling function (if it exists). 

How do you read the specification; do you have additional information?

**Reply by scruss** [reply]
> financial support for micropython

Espressif are (or were) project [sponsors](https://github.com/sponsors/micropython).

**Reply by mmliam** [reply]
That's interesting; I'm surprised that Espressif as a sponsor of micropython doesn't produce products with it already embedded.
I guess you don't have any further information regarding the ESP8266's ability to internally sample it's own supply voltage?

**Reply by scruss** [reply]
system_get_vdd33 on p.22 of https://www.espressif.com/sites/default/files/2c-esp8266_non_os_sdk_api_guide_en_v1.5.4.pdf

it's used in the ESP8266 MicroPython source here: [micropython/machine_adc.c at b41aaaa8a918a6645ebc6bfa4483bd17286f9263 · micropython/micropython](https://github.com/micropython/micropython/blob/b41aaaa8a918a6645ebc6bfa4483bd17286f9263/ports/esp8266/machine_adc.c#L46)

Do I know how to use it? No.
Have I seen lots of reports of it being unstable and inaccurate? Yes.

**Reply by mmliam** [reply]
You found something I missed the SDK API Guide, wow!
This is great.
Thank YOU!

**Comment by mmliam** [comment]
Thanks scruss for the information
